hikey.mk

```diff
@@ -133,26 +133,40 @@ busybox-cleaner: busybox-clean-common busybox-cleaner-common
 ################################################################################
 # EDK2 / Tianocore
 ################################################################################
-EDK2_VARS ?= EDK2_ARCH=AARCH64 \
-		EDK2_DSC=HisiPkg/HiKeyPkg/HiKey.dsc \
-		EDK2_TOOLCHAIN=GCC49 \
-		EDK2_BUILD=$(EDK2_BUILD)
+EDK2_ARCH ?= AARCH64
+EDK2_DSC ?= OpenPlatformPkg/Platforms/Hisilicon/HiKey/HiKey.dsc
+EDK2_TOOLCHAIN ?= GCC49
 
 EDK2_CONSOLE_UART ?= $(CFG_NW_CONSOLE_UART)
 ifeq ($(EDK2_CONSOLE_UART),0)
-	EDK2_VARS += EDK2_MACROS="-DSERIAL_BASE=0xF8015000"
+	EDK2_BUILDFLAGS += -DSERIAL_BASE=0xF8015000
 endif
 
 define edk2-call
-	GCC49_AARCH64_PREFIX=$(LEGACY_AARCH64_CROSS_COMPILE) \
-	$(MAKE) -j1 -C $(EDK2_PATH) \
-		-f HisiPkg/HiKeyPkg/Makefile $(EDK2_VARS)
+	$(EDK2_TOOLCHAIN)_$(EDK2_ARCH)_PREFIX=$(LEGACY_AARCH64_CROSS_COMPILE) \
+	build -n `getconf _NPROCESSORS_ONLN` -a $(EDK2_ARCH) \
```

Comment:
Not a big deal, but this `build` script (or function? I wasn't even able to find it in the edk2 sources!) will break proper parallel build. First problem: you are hard-coding _NPROCESSORS_ONLN, so it becomes impossible to build in a single thread (useful for troubleshooting). Second problem: you will likely end up with more build threads than specified on the command line. For instance `make -j2` should never run more than 2 subprocesses.

My suggestion: if there is a way to invoke $(MAKE) directly it's much better, i.e., simply remove the `-j1` and `make` will know the proper number of sub-processes it is allowed to spawn by communicating with the parent `make` process.

Reply:
Iiuc, `build` is actually `edk2/BaseTools/Source/Python/build/build.py`, so there's no way to invoke $(MAKE) directly.

There used to be a Makefile but not anymore. The Makefile for Arm Platform (`fvp.mk`, `qemu_v8.mk`) still exists in `edk2/ArmPlatformPkg/Scripts/Makefile`, but it's just a wrapper for `build` too, and so both these .mk files also hard-code `-n 6` in their `edk2-call`.

If we don't specify the number of threads, i.e not using `-n` option, and if I'm reading `build.py` and related code properly, it defaults to a single threaded build, which means it breaks proper parallel build too? Which means safer but slower?

In any case, I've tried various combinations of 1 and _NPROCESSORS_ONLN for both the `edk2` target and `edk2-call`, and this is the only one that never fails to build, and fast too! Even `-j1` in `edk2:` and `-n 1` in `edk2-call`, which used to be stable previously, failed for some reason.

Reply:
OK, I'm ready to accept that as a valid answer because I'm really not eager to go and try to understand the edk2 build system :O